django/contrib/auth/decorators.py

```diff
@@ -35,6 +35,35 @@ def _wrapper_view(request, *args, **kwargs):
 
             return redirect_to_login(path, resolved_login_url, redirect_field_name)
 
+        if asyncio.iscoroutinefunction(view_func):
+
+            @wraps(view_func)
+            async def _wrapper_view(request, *args, **kwargs):
+                if asyncio.iscoroutinefunction(test_func):
+                    result = await test_func(await request.auser())
+                else:
+                    try:
+                        result = test_func(await request.auser())
+                    except SynchronousOnlyOperation:
```

Comment:
Why do we need to catch `SynchronousOnlyOperation`? This feels wrong to me

I think the better implementation is to just use the code from the `except` block instead of calling it from an async context first. That method makes this work the same as the sync `_wrapper_view` below and would be more like what users would expect to happen.

Reply:
That may be true, there is a perf cost to making it work that way but it does make the implementation more consistent. My concern is that blindly calling a function twice is going to lead to more problems (and worse perf) than deterministic functionality.

I think we can safely trade perf for consistency here, as the perf notes are well documented: https://docs.djangoproject.com/en/5.0/topics/async/#performance

One option might be raising a warning when we encounter "mixed decoration" states (wrapper is async and view is sync, or the other way around). What do you think about that? AIUI we have a warning for when Middleware must be "adapted" to different calling environments so there is precedent for this.

Reply:
Oh, you are absolutely right, consistent is more important. I'll update the code later, thank you.

Reply:
I thought to raise a warning, when encounter "mixed decoration" states. But sometimes, the test_func doesn't contain any IO indeed, so it's enough to be sync (for example, the test_func within login_required).I thought to raise a warning, when encounter "mixed decoration" states. But sometimes, the test_func doesn't contain any IO indeed, so it's enough to be sync (for example, the test_func within login_required).

Reply:
> I thought to raise a warning when encounter "mixed decoration" states. But sometimes, the test_func doesn't contain any IO indeed, so sync is enough

Great point, I like the current implementation. Thanks for working through alternatives!